File: core/src/client/AuthHandle.ts
import type { LogoutReason, TokenSet } from '../types.js';
import type { MorphRuntime } from '../runtime.js';
import { listAuthIdsForProvider } from '../config/validate.js';

export class AuthHandle {
  constructor(
    private readonly rt: MorphRuntime,
    readonly authId: string,
  ) {}

  async submitCode(code: string, opts?: { codeVerifier?: string; redirectUriOverride?: string }): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    if (r.kind !== 'context') throw new Error('submitCode requires a provider/context auth id');
    this.rt.assertAlive();
    await this.rt.tokens.submitCode(r.authId, r.ref, code, opts);
  }

  async acquireWithClientCredentials(): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    if (r.kind !== 'context') throw new Error('acquireWithClientCredentials requires a provider/context auth id');
    this.rt.assertAlive();
    await this.rt.tokens.acquireWithClientCredentials(r.authId, r.ref);
  }

  async exchangeToken(targetAuthId: string): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    if (r.kind !== 'context') throw new Error('exchangeToken requires a provider/context auth id as the source');
    this.rt.assertAlive();
    await this.rt.tokens.exchangeToken(r.authId, r.ref, targetAuthId);
  }

  async setTokens(tokens: TokenSet): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    if (r.kind !== 'context') throw new Error('setTokens requires a provider/context auth id');
    this.rt.assertAlive();
    await this.rt.tokens.setTokens(r.authId, r.ref, tokens);
  }

  async clearTokens(): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    this.rt.assertAlive();
    if (r.kind === 'context') {
      await this.rt.tokens.clearTokens(r.authId, r.ref);
      return;
    }
    const ids = listAuthIdsForProvider(r.providerKey, this.rt.resolved);
    for (const id of ids) {
      const ref = this.rt.resolved.contextByAuthId.get(id)!;
      await this.rt.tokens.clearTokens(id, ref);
    }
  }

  async logout(reason: LogoutReason = 'user_initiated'): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    this.rt.assertAlive();
    if (r.kind === 'context') {
      await this.rt.tokens.logout(r.authId, r.ref, reason);
      return;
    }
    await this.rt.tokens.logoutProvider(r.providerKey, reason);
  }

  async hasValidToken(): Promise<boolean> {
    const r = this.rt.parseAuthRef(this.authId);
    this.rt.assertAlive();
    if (r.kind === 'context') return this.rt.tokens.hasValidTokenContext(r.authId, r.ref);
    return this.rt.tokens.hasValidTokenProvider(r.providerKey);
  }

  async refreshTokens(): Promise<void> {
    const r = this.rt.parseAuthRef(this.authId);
    if (r.kind !== 'context') throw new Error('refreshTokens requires a provider/context auth id');
    this.rt.assertAlive();
    await this.rt.tokens.refreshTokensManual(r.authId, r.ref);
  }

  async peekTokens(): Promise<TokenSet | null> {
    const r = this.rt.parseAuthRef(this.authId);
    if (r.kind !== 'context') throw new Error('peekTokens requires a provider/context auth id');
    this.rt.assertAlive();
    return this.rt.tokens.loadTokens(r.authId, r.ref);
  }
}

File: packages/dart/morph_core/lib/src/client/auth_handle.dart
import '../config/list_auth_ids.dart';
import '../runtime/morph_runtime.dart';
import '../types/morph_surface.dart';
import '../util/jwt_utils.dart';

/// Parity: [`AuthHandle`](packages/ts/core/src/client/AuthHandle.ts).
final class AuthHandle {
  AuthHandle(this._rt, this.authId);

  final MorphRuntime _rt;
  final String authId;

  Future<void> submitCode(String code, {String? codeVerifier, String? redirectUriOverride}) async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('submitCode requires a provider/context auth id');
    }
    _rt.assertAlive();
    await _rt.tokens.submitCode(r.authId, r.ref, code,
        codeVerifier: codeVerifier, redirectUriOverride: redirectUriOverride);
  }

  Future<void> acquireWithClientCredentials() async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('acquireWithClientCredentials requires a provider/context auth id');
    }
    _rt.assertAlive();
    await _rt.tokens.acquireWithClientCredentials(r.authId, r.ref);
  }

  Future<void> exchangeToken(String targetAuthId) async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('exchangeToken requires a provider/context auth id as the source');
    }
    _rt.assertAlive();
    await _rt.tokens.exchangeToken(r.authId, r.ref, targetAuthId);
  }

  Future<void> setTokens(TokenSet tokens) async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('setTokens requires a provider/context auth id');
    }
    _rt.assertAlive();
    await _rt.tokens.setTokens(r.authId, r.ref, tokens);
  }

  Future<void> clearTokens() async {
    final r = _rt.parseAuthRef(authId);
    _rt.assertAlive();
    if (r is ParsedAuthContext) {
      await _rt.tokens.clearTokens(r.authId, r.ref);
      return;
    }
    final pk = r is ParsedAuthProvider ? r.providerKey : '';
    for (final id in listAuthIdsForProvider(pk, _rt.resolved)) {
      final ref = _rt.resolved.contextByAuthId[id];
      if (ref != null) await _rt.tokens.clearTokens(id, ref);
    }
  }

  Future<void> logout([String reason = 'user_initiated']) async {
    final r = _rt.parseAuthRef(authId);
    _rt.assertAlive();
    if (r is ParsedAuthContext) {
      await _rt.tokens.logout(r.authId, r.ref, reason);
      return;
    }
    if (r is ParsedAuthProvider) {
      await _rt.tokens.logoutProvider(r.providerKey, reason);
    }
  }

  Future<bool> hasValidToken() async {
    final r = _rt.parseAuthRef(authId);
    _rt.assertAlive();
    if (r is ParsedAuthContext) {
      return _rt.tokens.hasValidTokenContext(r.authId, r.ref);
    }
    if (r is ParsedAuthProvider) {
      return _rt.tokens.hasValidTokenProvider(r.providerKey);
    }
    return false;
  }

  Future<void> refreshTokens() async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('refreshTokens requires a provider/context auth id');
    }
    _rt.assertAlive();
    await _rt.tokens.refreshTokensManual(r.authId, r.ref);
  }

  Future<TokenSet?> peekTokens() async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('peekTokens requires a provider/context auth id');
    }
    _rt.assertAlive();
    return _rt.tokens.loadTokens(r.authId, r.ref);
  }

  /// Decoded access token JWT claims, or null when no token / opaque token. No network, no refresh.
  Future<JwtPayload?> getClaims() async {
    final r = _rt.parseAuthRef(authId);
    if (r is! ParsedAuthContext) {
      throw StateError('getClaims requires a provider/context auth id');
    }
    _rt.assertAlive();
    if ((r.ref.context.tokenTypes['access']?.format ?? 'jwt') == 'opaque') {
      return null;
    }
    final set = await _rt.tokens.loadTokens(r.authId, r.ref);
    if (set?.accessToken == null || set!.accessToken.isEmpty) return null;
    try {
      return decodeJwtPayload(set.accessToken);
    } catch (_) {
      return null;
    }
  }
}
